python/src/nanoarrow/c_array.py

```diff
@@ -0,0 +1,573 @@
+    - null_count: 0
+    - buffers[2]:
+      - validity <bool[0 b] >
+      - data <uint8[5 b] 49 50 51 52 53>
+    - dictionary: NULL
+    - children[0]:
+    """
+    schema = c_schema(schema)
+    builder = CArrayBuilder.allocate()
+
+    # Ensures that the output array->n_buffers is set and that the correct number
+    # of children have been initialized.
+    builder.init_from_schema(schema)
+
+    # Set buffers, optionally moving ownership of the buffers as well (i.e.,
+    # the objects in the input buffers would be replaced with an empty ArrowBuffer)
+    for i, buffer in enumerate(buffers):
+        if buffer is None:
+            continue
+
+        # If we're setting a CBuffer from something else, we can avoid an extra
+        # level of Python wrapping by using move=True
+        move = move or not isinstance(buffer, CBuffer)
+        builder.set_buffer(i, c_buffer(buffer), move=move)
+
+    # Set children, optionally moving ownership of the children as well (i.e.,
+    # the objects in the input children would be marked released).
+    n_children = 0
+    for child_src in children:
+        # If we're setting a CArray from something else, we can avoid an extra
+        # level of Python wrapping by using move=True
+        move = move or not isinstance(child_src, CArray)
+        builder.set_child(n_children, c_array(child_src), move=move)
+        n_children += 1
+
+    if n_children != schema.n_children:
+        raise ValueError(f"Expected {schema.n_children} children but got {n_children}")
+
+    # Set array fields
+    builder.set_length(length)
+    builder.set_offset(offset)
+    builder.set_null_count(null_count)
+
+    # Calculates the null count if -1 (and if applicable)
+    builder.resolve_null_count()
+
+    # Validate + finish
+    return builder.finish(validation_level=validation_level)
+
+
+class ArrayBuilder:
+    """Internal utility to build CArrays from various types of input
+
+    This class and its subclasses are designed to help separate the code
+    that actually builds a CArray from the code that chooses the strategy
+    used to do the building.
+    """
+
+    @classmethod
+    def infer_schema(cls, obj) -> Tuple[CSchema, Any]:
+        """Infer the Arrow data type from a target object
+
+        Returns the type as a :class:`CSchema` and an object that can be
+        consumed in the same way by append() in the event it had to be
+        modified to infer its type (e.g., for an iterable, it would be
+        necessary to consume the first element from the original iterator).
+        """
+        raise NotImplementedError()
+
+    def __init__(self, schema):
+        self._schema = c_schema(schema)
+        self._schema_view = c_schema_view(self._schema)
+
+    def build_c_array(self, obj):
+        builder = CArrayBuilder.allocate()
+        builder.init_from_schema(self._schema)
+        self.start_building(builder)
+        self.append(builder, obj)
+        return self.finish_building(builder)
```

Comment:
Instead of having to pass the c_builder around to each method, could this be part of the object state?